### ISSUE TITLE
Play 2.4 support

### ISSUE THREAD

**Issue by francisdb** [ReactiveMongo/Play-ReactiveMongo#125]
Since play framework 2.4 is released switching master to it seems like a good idea?

Issues I currently have:
```scala
type arguments [play.modules.reactivemongo.json.collection.JSONCollection] do not conform to method collection's type parameter bounds [C <: reactivemongo.api.Collection]
[error]           mongo.db.collection[JSONCollection](name).indexesManage
```
and
```scala
could not find implicit value for parameter swriter: GenericCollection.this.pack.Writer[play.api.libs.json.JsObject]
```

**Comment by RavisMsk** [comment]
Also this problem seem to be caused by Play 2.4-RC5: https://github.com/ReactiveMongo/Play-ReactiveMongo/issues/124

**Comment by francisdb** [comment]
I can confirm I am using `Play 2.4-RC5`

**Comment by RavisMsk** [comment]
@francisdb as i understood from sources, the problem you posted is somehow connected to on-going re-factoring of this module.
I've downloaded the sources, and i couldn't successfully `publish-local`.
At the moment I'm trying to make a working fork of this module for Play2.4. I decided to down-grade main module ReactiveMongo to version 0.10.5.0-akka23, so this module is compatible with ReactiveMongo.
Also i've updated Play to 2.4.0-RC5, and fixed all the incompatibilities with changed JSON api. 
But i've met a new problem with log4j:
```
java.lang.AbstractMethodError: org.apache.logging.slf4j.SLF4JLoggerContextFactory.getContext(Ljava/lang/String;Ljava/lang/ClassLoader;Ljava/lang/Object;Z)Lorg/apache/logging/log4j/spi/LoggerContext;
	at org.apache.logging.log4j.LogManager.getLogger(LogManager.java:474)
	at reactivemongo.utils.LazyLogger$.apply(utils.scala:55)
	at reactivemongo.core.actors.MongoDBSystem$.<init>(actors.scala:604)
	at reactivemongo.core.actors.MongoDBSystem$.<clinit>(actors.scala)
	at reactivemongo.api.MongoDriver$$anonfun$4.apply(api.scala:382)
	at reactivemongo.api.MongoDriver$$anonfun$4.apply(api.scala:382)
	at akka.actor.TypedCreatorFunctionConsumer.produce(Props.scala:336)
	at akka.actor.Props.newActor(Props.scala:256)
        ...
```
And i can't find the solution at the moment.
You can find fork at [my profile](https://github.com/RavisMsk/Play-ReactiveMongo/tree/play24-compatibility).

I would appreciate any help!

Thanks in advance!

**Comment by francisdb** [comment]
2.4 has been released...

**Comment by francisdb** [comment]
@sgodbillon are you still maintaining the project?

**Comment by RavisMsk** [comment]
@francisdb i've made a fast work-around for Play2.4-RC5. ATM i'm using it in my project, seems to work pretty well. Check it out - https://github.com/ReactiveMongo/Play-ReactiveMongo/pull/126.
Seems like this project is not actively maintained at the moment...

**Comment by francisdb** [comment]
We can't deploy anything not available in maven central so this is currently holding us back from switching to play 2.4

**Comment by ticofab** [comment]
+1

**Comment by bmontuelle** [comment]
+1 , we are willing to contribute some code / testing if needed

**Comment by francisdb** [comment]
We'll have to wait for a response from @sgodbillon ?

**Comment by fhessenberger** [comment]
for some reason, (maybe because of JSON API changes) the reads for all special mongo subdocuments ($oid, $date, etc.) failed to match, resulting in raw JSON data being sent to the DB (instead of ObjectId("..")) (json.scala)

```scala
implicit object BSONObjectIDFormat extends PartialFormat[BSONObjectID] {
  def partialReads: PartialFunction[JsValue, JsResult[BSONObjectID]] = {
    case JsObject(Seq(("$oid", JsString(v)))) => JsSuccess(BSONObjectID(v))
  }
  val partialWrites: PartialFunction[BSONValue, JsValue] = {
    case oid: BSONObjectID => Json.obj("$oid" -> oid.stringify)
  }
}
```

the solution was to rewrite the matching a little bit (i did this via an unapply function).

```scala
implicit object BSONObjectIDFormat extends PartialFormat[BSONObjectID] {
  object JsObjectId {
    def unapply(js: JsValue): Option[String] = (js \ "$oid").asOpt[String]
  }
  def partialReads: PartialFunction[JsValue, JsResult[BSONObjectID]] = {
    case JsObjectId(s) => JsSuccess(BSONObjectID(s))
  }
  val partialWrites: PartialFunction[BSONValue, JsValue] = {
    case oid: BSONObjectID => Json.obj("$oid" -> oid.stringify)
  }
}
```

if someone else has this problem, please consider my solution (i'm not actively contributing to this repo and also it seems the author is not maintaining this (great) project anymore)

**Comment by RadoBuransky** [comment]
I have just created a pull request where these issues seem to be fixed:
https://github.com/ReactiveMongo/Play-ReactiveMongo/pull/129

The plugin is deprecated from now and you should use module and API. More info here:
https://github.com/RadoBuransky/Play-ReactiveMongo/tree/play240

The only problem is that it is not yet availabe in any public repository you have to build it by yourselves by doing `sbt publish-local`.

**Comment by fhessenberger** [comment]
:+1:

**Comment by RavisMsk** [comment]
@RadoBuransky Look #126, already fixed and sent pull request a week ago.

**Comment by RadoBuransky** [comment]
@RavisMsk Your pull request is for RC5 and it didn't work for me with Play 2.4. Also I have added support for depndency injection and deprecated the static plugin object. Anyway, I hope the project is not dead and that somebody will merge it and release it.

**Comment by francisdb** [comment]
:+1:

**Comment by RavisMsk** [comment]
While pull request is not merged and new is not released on maven central, there is a workaround for people trying to deploy apps on PaaS (meaning that you cant first clone fork, then `publish-local`).
1) Clone it on your local, then perform `package`. This will create jar w/o dependencies;
2) Move jar to `/lib/` folder of your project;
3) Manually include all deps of included lib into your `build.sbt`.
Then push it on any PaaS. Currently, i'm testing that on Heroku, should work until officially released.

**Comment by francisdb** [comment]
you can also point to a git repo as a dependency and sbt will take care of building the project
https://www.safaribooksonline.com/library/view/scala-cookbook/9781449340292/ch18s11.html

**Comment by RavisMsk** [comment]
@francisdb tried that method, but couldn't make it work. How did you manage to run that?

**Comment by francisdb** [comment]
@RavisMsk have not tried this for this specific project but previously used it for other projects.

**Comment by andrey-feoktistov** [comment]
@francisdb I use this in my project
```
  lazy val playReactiveMongo = RootProject(uri("git://github.com/RadoBuransky/Play-ReactiveMongo.git#play240"))
```

**Comment by RadoBuransky** [comment]
@andrey-feoktistov What dependencies are wrong? "It works on my machine"

**Comment by andrey-feoktistov** [comment]
@RadoBuransky Hi! Thank you for you work! I have dependency issue when I try to do test:compile or add the git project as the dependency, seems like the sbt trying to download all dependencies including required only for tests in this case. 
```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: org.scalaz.stream#scalaz-stream_2.11;0.7a: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]          org.scalaz.stream:scalaz-stream_2.11:0.7a
[warn]            +- org.specs2:specs2-common_2.11:3.6.1
[warn]            +- org.specs2:specs2-matcher_2.11:3.6.1
[warn]            +- org.specs2:specs2-core_2.11:3.6.1 (/home/cattus/projects/satrunner/Play-ReactiveMongo/project/Play2-ReactiveMongo.scala#L125)
[warn]            +- org.reactivemongo:play2-reactivemongo_2.11:0.10.5.0.akka23.play24-SNAPSHOT
sbt.ResolveException: unresolved dependency: org.scalaz.stream#scalaz-stream_2.11;0.7a: not found
```

so the solution for me was to add ```"scalaz-bintray" at "http://dl.bintray.com/scalaz/releases"```.

**Comment by cchantep** [comment]
There are PR waiting for the Play 2.4 support. There will be a review campaign this week to go forward. Thanks to all the contributors. Best regards.

**Comment by francisdb** [comment]
:+1: big thanks @cchantep

**Comment by carlosFattor** [comment]
It's good listen this... Tanks

**Comment by cchantep** [comment]
Some PR have already been merged on the [driver repo](https://github.com/ReactiveMongo/Play-ReactiveMongo). The review process is going on, and the PR on for the Play plugin are the next on the list. Best regards.

**Comment by RollandWaters** [comment]
Andrey's solution worked for me, but for those of us just learning sbt, the full line you want to add to build.sbt is:

resolvers += "scalaz-bintray" at "http://dl.bintray.com/scalaz/releases".

**Comment by dragoninfrontofme** [comment]
Hi guys, is play-reactivemongo support play 2.4 now? i am plan to upgrade my play framework but considering this plugin, because currently im using it in my production :)

**Comment by francisdb** [comment]
Nope, I suggest you wait for a release when @cchantep is done reviewing and merging the pull request(s)

**Comment by dragoninfrontofme** [comment]
@francisdb well noted, thanks

**Comment by mrkaspa** [comment]
Is the support for 2.4 ready or is there another solution?

**Comment by garrrio** [comment]
@mrkaspa please see https://github.com/ReactiveMongo/Play-ReactiveMongo/pull/129

**Comment by cchantep** [comment]
@mrkaspa @garrrio better to see #131 which is published as `"org.reactivemongo" %% "play2-reactivemongo" % "0.11.0.play24-SNAPSHOT"` on OSS Sonatype repository, and will support both the 0.11.0 driver & Play 2.4.

**Comment by mrkaspa** [comment]
@cchantep can I use that with the actual reactive mongo extensions or how can I use both, because I tryed the @garrrio solution but  I get a dependency error with the reactive mongo extensions

**Comment by mrkaspa** [comment]
@cchantep it's not on OSS or I couldn't find it

**Comment by cchantep** [comment]
The version `0.11.0.play23-SNAPSHOT` is available on OSS Snapshots ( https://oss.sonatype.org/content/repositories/snapshots/org/reactivemongo/play2-reactivemongo_2.11/0.11.0.play23-SNAPSHOT/ ).

If you have more question, please first ask one the [user MailingList](https://groups.google.com/forum/m/?fromgroups#!forum/reactivemongo), rather on this issue tracker.

**Comment by francisdb** [comment]
@cchantep you might want to document that in the readme

**Comment by cchantep** [comment]
The README will be updated when the plugin is published.

**Comment by francisdb** [comment]
There is no mention of the forum and it has outdated snapshot instructions

**Comment by cchantep** [comment]
@francisdb The forum is mentionned in the contributor guidelines.

**Comment by timprobert** [comment]
@cchantep Hi Cedric,
I've just followed the readme though I can't resolve your plugin. Is it fully up yet?

 unresolved dependency: org.reactivemongo#play2-reactivemongo_2.11;0.11.0.play24: not found

**Comment by francisdb** [comment]
Hmm, has the final been released? Maven central currently only has the M2 version
https://repo1.maven.org/maven2/org/reactivemongo/play2-reactivemongo_2.11/

**Comment by timprobert** [comment]
Not yet it seems. I think I jumped the gun. Play 2.4 support was merged an hour or so ago whereas M2 was updated yesterday. I'm just a little keen. I'll learn to wait a bit longer :)

**Comment by cchantep** [comment]
Merged branch is the one published as M2. It's merged now as it's about to be published as final 0.11.
